### ISSUE TITLE
Allow users to determine package resolution strategy during package restore - direct or transitive

### ISSUE THREAD

**Issue by Wil73** [NuGet/Home#5553]
I am not going to debate if highest or lowest should be the default: leave the actual behavior up to the consumer.

Why on earth should Microsoft FORCE users to adopt the "safest is best" strategy in the first place?

Is there any reason why we cannot add a switch to nuget.exe and NPM that allows the user to decide their own package restore strategy?

This way consumers of nuget can align nuget behavior to their own business strategies, be they conservative or aggressive.

What I don't understand is why Microsoft has to decide this for consumers in the first place.

To me this is a straight up consumer decision based on private business strategy and policy.

It makes absolutely no sense why a tool vendor should allow themselves to unilaterally decide business strategy for all nuget consumers on the planet without an option to override that.

It's been two years and there's no apparent movement on this issue.

Leave lowest version as the default strategy if you want Microsoft, but for goodness sakes offer a switch to allow consumers to make their own business decisions.

**Comment by davidfowl** [comment]
> I am not going to debate if highest or lowest should be the default: leave the actual behavior up to the consumer.

Lowest should be the default, but that's just my opinion. It's the best way to get a "working set of packages". Everyone in the graph gets what they compiled against and things only roll forward minimally if sibling packages need higher versions.

> Why on earth should Microsoft FORCE users to adopt the "safest is best" strategy in the first place?

It's more of a "safest is default strategy" because people want to get a working set of packages by default. In a system where you *need* binary compatibility between components to in order to make things function things have a better chance of working if you stay closer to what components build, and test against.

> Is there any reason why we cannot add a switch to nuget.exe and NPM that allows the user to decide their own package restore strategy?

This is already possible when using packages.config based projects by picking the dependency behavior on install:

![image](https://user-images.githubusercontent.com/95136/27986627-2694d382-63b6-11e7-8174-3d4c21fd8a21.png)

We don't own NPM, that's the node package manager, not sure if you mean something else.

To properly do this with PackageReference, a lock file needs to be introduced (which has been discussed anyways).

**Comment by Wil73** [comment]
I am not going to debate what the default ought to be.  I really am not concerned over that.  Technically this is NOT a default when you can't configure it otherwise. It's only a default if you can change it.  We can't.

My company is frustrated that Microsoft will not allow businesses to DECIDE Nuget resolution behavior for themselves.

It makes absolutely no sense why Microsoft should have a monopoly on the business policies of the private sector: this is in effect what Microsoft does here.

Make package resolution configurable via a simple switch.

By NPM I mean NuGet Package manager in Visual Studio.

I don't see why any lock file is required.  Just add an additional switch in nuget.exe and Nuget Package Manager that indicates if lowest, highest etc should be the expected behavior.

Package.Config is for NuGet 1 and 2. That's well deprecated by now.

This is clearly a business decision that Microsoft currently has constrained to Microsoft to decide.

There are many switches in Nuget already.  I don't understand why we cannot simply add one more.

I fail to understand why it takes over two years for Microsoft to resolve this situation.  The current work around is to explicitly declare EVERY dependency in package references (for Nuget 4) and in Project.Json (for Nuget 3).  This is a clumsy and labor intensive work around.

Just add a darn switch to indicate Nuget's expected resolution behavior so users can manage it themselves.

It makes no sense why Microsoft must control this in the first place.

Just add a switch.

**Comment by davidfowl** [comment]
> I don't see why any lock file is required. Just add an additional switch in nuget.exe and Nuget Package Manager that indicates if lowest, highest etc should be the expected behavior.

It's required with `PackageReference` because there's no way to make restore deterministic if you don't add it. If there's no asset you can commit to source control to lock dependencies then restoring the latest versions of all dependencies could give dramatically different results build to build. That might not be an issue if you tightly control the feeds but it is in the general case so just adding a switch isn't sufficient for the masses. This feature needs to be paired with a lock file to make it sane to use for the majority of customers.

**Comment by Wil73** [comment]
I think having the flexibility to choose resolution strategies is just as important as having a deterministic build.  I do agree a deterministic build is important, but so is flexibility, which has been in short supply for long enough.  

There is so much debate as to what the restore default should be (highest or lowest), but the point is irrelevant: higher or lower should be a choice the business makes, not Microsoft.

If I want a predictable build I can always just stick with project.json, check in my project.lock.json file with lock set to true, or just use fixed version numbers (the latter is the most predictable way, but the least flexible and most labor intensive).  A package can get de-listed at any moment, so defaulting to lowest version does not guarantee a predictable build output either.

Microsoft needs to add switches to allow business to define their own resolution defaults.  Microsoft has no business making these decisions for businesses in the private sector.

**Comment by davidfowl** [comment]
> There is so much debate as to what the restore default should be (highest or lowest), but the point is irrelevant: higher or lower should be a choice the business makes, not Microsoft.

Sure. Like I said, it's possible with packages.config and that functionality should be ported to PackageReference (project.json is pretty much deprecated at this point) along with a way to lock the graph after resolution.

> I think having the flexibility to choose resolution strategies is just as important as having a deterministic build. I do agree a deterministic build is important, but so is flexibility, which has been in short supply for long enough.

I think you can have both though. IMO it wouldn't be great if we had a feature that most people couldn't use because we didn't think through the end to end.

> If I want a predictable build I can always just stick with project.json, check in my project.lock.json file with lock set to true, 

That doesn't work. Locked: true isn't supported anymore (hasn't been for a while).

> or just use fixed version numbers (the latter is the most predictable way, but the least flexible and most labor intensive). 

Fixed versions of the entire dependency graph would work, but it's extremely labor intensive.

> A package can get de-listed at any moment, so defaulting to lowest version does not guarantee a predictable build output either.

I believe this isn't respected when resolving dependencies in `project.json` so it *is* deterministic.

**Comment by Wil73** [comment]
project.json will resolve to the lowest version it can find.  If the version the build uses normally gets de-listed your deterministic build is out the window.

If you are going to use dependencies from nuget.org or any other third party site, you cannot get a deterministic build every time, even with nuget restore working as it does now.

When you depend on a third party for some of your libraries you cannot determine what your build will look like every time no.

The alternative is to host any libraries you use locally so you have complete control over their list status.

I think we both agree package.config with fixed version numbers is far too labor intensive.

Project.json is far more flexible than package.config.  Package.config is even more deprecated and out of date than project.json.

We still use project.json where I work along with Nuget 3.5 since Nuget 4 does not offer any more flexibility with regards to package resolution options.

When Microsoft smartens up and lets go of its monopoly on package resolution strategy logic we will adopt a later version.

As it stands now Nuget 4 does not offer anything that Nuget 3.5 doesn't already do for us.  Locked: true does work for us with Nuget 3.5.  Though we don't want fixed version numbers.  We can update a library quite regularly.  We need flexibility here more than we need predictability.

We use TFS automated builds with a fully featured Artifact Server here as well.  If we need any build of any configuration we can pull that from TFS in just seconds.  Flexible builds are a more pressing need for my company than deterministic builds.

The Lower version is no guarantee of safer.  If that was the case we would all still be running Windows 95.

**Comment by davidfowl** [comment]
> project.json will resolve to the lowest version it can find. If the version the build uses normally gets de-listed your deterministic build is out the window.

1. It's *super* uncommon to de-list things. That said, I'm not even sure restore respects this.
2. If you delete things, your build is broken anyways.

> If you are going to use dependencies from nuget.org or any other third party site, you cannot get a deterministic build every time, even with nuget restore working as it does now.

You can 98% of the time (I made that up), because de-listing on nuget.org is rare and deleting is impossible. Now, if you're mirroring your dependencies to a myget feed or controlled feed, it's even better since you'd rarely delete or de-list anything that was being depended on. 

However, de-listing/deleting is rare, if we support resolving the latest dependency version without a sane story to lock it, we're turning the 2% case into the 98% case.

> When you depend on a third party for some of your libraries you cannot determine what your build will look like every time no.

Not following this argument.

> Project.json is far more flexible than package.config. Package.config is even more deprecated and out of date than project.json.

> We still use project.json where I work along with Nuget 3.5 since Nuget 4 does not offer any more flexibility with regards to package resolution options.

Hopefully you'll switch to PackageReference since that's where all of our energy is at the moment. Good to know `locked: true` still works in the version of the client that you're using.

All of that said I agree with you, I just think when we do that feature it'll be accompanied with a lock file so that things remain deterministic. That shouldn't be too hard to accomplish regardless.

**Comment by Wil73** [comment]
I would have to double check for Nuget 4, but you can get a deterministic build using highest version with a version number limit.  I don't think Nuget 4 will allow this scenario out of the box either.  It's not 100% predictable, but the point is we cannot even configure that.

If you want to lock, then use VS 2015 and Nuget 3.5.  You just have to check in your project.lock.json. 
 That works just fine.  I haven't investigated locking for Nuget 4 and VS 2017, but it sounds like Microsoft dropped the ball yet again.

I like the high level design of NuGet, but the implementation is dreadful on many levels.  MS has had over five years now to get NuGet to a sensible level and it's still not there.

If locks are out the window with Nuget 4, and Nuget 4 STILL won't offer a switch for businesses to configure their own business policies how is Nuget 4 any better?

Moving references to the project file accomplishes what?  That work effort could have been put into offering what businesses need: configurable restore policies and restore locks.  I don't recall seeing a single post of anyone complaining that the project.json file format urgently needs a renovation.

There are hundreds (I haven't counted, but I stopped a long time ago) of people complaining about the inflexible lowest is best logic hard coded into Nuget right now.

On the face of it, it looks like MS wants everyone to use the latest version of their software, but the oldest version of everyone else's software.  I find it ironic how aggressively Microsoft endorses adoption of new releases of their products, but Nuget logic has a "lowest is safer" constraint hard coded into the binary.

If lowest was safer we would all still be running Windows 95.

**Comment by davidfowl** [comment]
> I would have to double check for Nuget 4, but you can get a deterministic build using highest version with a version number limit. I don't think Nuget 4 will allow this scenario out of the box either. It's not 100% predictable, but the point is we cannot even configure that.

Not only is it not predictable, it's probably more broken (especially for dependencies out of your control).

> If you want to lock, then use VS 2015 and Nuget 3.5. You just have to check in your project.lock.json.
That works just fine. 

VS 2017 with PackageReference is the future of nuget and it needs a proper lock file story.

**Comment by Wil73** [comment]
How is getting the latest version "out of control?"

Our approach at our company is to use a fixed version number for third party dependencies and use the highest version for in house packages.  We don't update third party dependencies nearly as often as we update internal packages.  We have a very large in house class library framework that relies on nuget to be sensible: right now it's not.  We can make nuget sensible, but we have to hack project.json to make that happen.

This is not out of control.  Our dependencies are quite stable.  We have a solid change management process and automated testing.  Adopting a new version when that version gets released does not make a process out of control.  It just makes your company agile and aggressive in adopting the new.

This goes back to my original point: this is a controversial debate left to the business to decide, not a tool vendor.

I cannot say for any given company if lower or higher version resolution strategy is better. I think it depends on the company and how aggressively they want to adopt new tools.  Again this is such a controversial topic it is absurd to have a tool vendor make these decisions for everyone and hard code those policies into their binaries.

That being said, just offer a darn switch.  That ends all debate on the matter. lol

**Comment by davidfowl** [comment]
> How is getting the latest version "out of control?"

You misread, I said "dependencies that are out of your control".

> We don't update third party dependencies nearly as often as we update internal packages.

Sure. How do you restrict the resolution strategy to only update certain dependencies?

> This goes back to my original point: this is a controversial debate left to the business to decide, not a tool vendor.

I think I agreed with you each time you made this comment but my position hasn't changed when it comes to needing a lock file with this feature.

> That being said, just offer a darn switch. That ends all debate on the matter. lol

I have little say in what the nuget team does here so I can't say how or when this would ever happen. I'm guessing it won't work on VS 2015 with `project.json` though (I could be wrong). I'm also guessing this mode needs to be source controllable (probably in nuget.config) since you'd want command line restore and  VS restore to do the same thing.

**Comment by Wil73** [comment]
You can use project.json syntax to restrict to specific versions or simply de-list other packages so you only have one version in your package line.  It's a hack, but it works.

Yes, you would want a switch or option for both NPM and the command line restore of course.

**Comment by anangaur** [comment]
That's a long thread over the weekend. :)  
From what I have observed (here - this thread as well as talking to other folks), the requirements seem to be the following:
1. Allow users to decide the dependency resolution strategy - direct as well as transitive 
2. Allow users to lock down on the dependency versions in the transitive world - to enable repeatable build

Both # 1 and # 2 seem to be lacking with PackageReference. # 2 seems like already working with lowest is the best strategy but will falter with # 1 coming into the fore. So I think NuGet will have to solve both together atleast for PackageReference.

**Comment by giggio** [comment]
What is the timeline to resolve this? This is very much needed. Requests for this goes back to 2015 (see #4789 and https://github.com/aspnet/dnx/issues/2657).

**Comment by Wil73** [comment]
I think they just closed it and ignored me.  And yes, MANY people have raised this as a SERIOUS flaw in the Nuget system.  It's been over five years and it still behaves the same way with no flexibility.

I have seen MANY tickets opened for this and they just close them and walk away.

**Comment by davidfowl** [comment]
> I think they just closed it and ignored me.

Closed what? The issue is still open. Like I said before, a lock file is needed before any realistic progress can be made here. 

> And yes, MANY people have raised this as a SERIOUS flaw in the Nuget system. It's been over five years and it still behaves the same way with no flexibility.

It has been raised many times but honestly when you talk through the implications, people ignore the consequences that don't affect their immediate needs. We need to think about the larger ecosystem impact when we design features like this. 

As a super basic simple example, any package that transitively references Newtonsoft.Json say 6.0.8, will now start pulling in 10.0.1. Not only that, but you end up with a completely jagged untested dependency graph by default. Here's an example of that going bad:

https://orientman.wordpress.com/2017/08/22/how-_not_-to-upgrade-to-asp-net-core-2-0-just-yet-with-paket/

I also think @anangaur wrote a spec for a proposed solution to the lock file but I can't find it on the wiki.

**Comment by anangaur** [comment]
@Wil73 We closed many related issues that were essentially the same ask in this issue. This issue is still open :)

Here is the "[Enable repeatable builds via lock file](https://github.com/NuGet/Home/wiki/Enable-repeatable-builds-for-PackageReference-based-projects)" spec.

**Comment by Wil73** [comment]
I have yet to grasp what this has to do with adding a switch of some sort to nuget RESTORE to ensure that packages resolve by default to the HIGHEST version available?

**Comment by Wil73** [comment]
I fail to see how on earth a lock file has any relevance at all here either.

**Comment by anangaur** [comment]
We already discussed this in the earlier [comments](https://github.com/NuGet/Home/issues/5553#issuecomment-314024910). We would need to handle both scenarios together.

**Comment by giggio** [comment]
Just use a default that maintains the current behavior, and allow for different expressions of version ranges. Npm does this brilliantly with the `^` and `~` characters, as I pointed out on #4789.
This way people that don't want to change the current behavior just keep doing what they do, and people who do want it use the new characters.
Don't get me wrong, I love the idea of a lockfile, but these are orthogonal needs and can be worked in parallel.

**Comment by Wil73** [comment]
Sorry, but that's just silly.  Every time you come up with a new version you then have to edit all the packages that use the new version to increase the darn range.

We have a * as an option right now, but that only works for direct dependencies... child dependencies still resolve to LOWEST version despite the * in the grandparent library.  Ranges will not resolve this problem; ranges will just create a maintenance nightmare.

**Comment by anangaur** [comment]
@giggio Enabling option for users to choose the resolution strategy is good but aggravates the "repeatable build" scenario. Something we do not want to do. Hence we would like to solve the both together (rather 'lock file' solution followed by 'user defined resolution strategy' <- this issue)

**Comment by Wil73** [comment]
Yes, I understand you want a build that's 100% predictable.  I do get it.  I just don't think that's as urgent as a flexible business strategy.  This problem has been in the air for years.  The workarounds are an administration nightmare.

If they can be resolved separately and quickly please explore that as an option first.

**Comment by davidfowl** [comment]
> Yes, I understand you want a build that's 100% predictable.

Not 100% any percent really. If you float every dependency every time restore is run it's extremely unpredictable unless you control *all* of the feeds.

> We have a * as an option right now, but that only works for direct dependencies... child dependencies still resolve to LOWEST version despite the * in the grandparent library. Ranges will not resolve this problem; ranges will just create a maintenance nightmare.

@emgarten do we support * in nuspec dependencies? I though that wasn't supported today. @Wil73 did you manually enter * in your nuspec and run nuget pack?

**Comment by Wil73** [comment]
We use the * in our project.json files and that works fine.  BUT we have to include EVERY dependency in the entire tree to ensure we get the most recent version.  That is the most common work around people use right now.  It's ugly and a maintenance nightmare, particularly for complex dependency trees.

We do not include any version in our nuspec files.  It doesn't matter so long as the application exe builds with a * for every nupkg he's after.

It's VERY easy to miss one since the application will build without it (silently pulling in OLDEST instead).

It's ugly and prone to error, but it works... for now.  Our program manager hates it.

**Comment by emgarten** [comment]
> @emgarten do we support * in nuspec dependencies? I though that wasn't supported today.

Floating versions in nuspecs are ignored.

**Comment by davidfowl** [comment]
@emgarten good 😄 .

> It's VERY easy to miss one since the application will build without it (silently pulling in OLDEST instead).

> That is the most common work around people use right now. It's ugly and a maintenance nightmare, particularly for complex dependency trees.

I really honestly don't understand why people would want the latest version of every dependency without the ability to build with those resolved versions more than once. The only reason this works in packages.config worked is because it resolved dependencies once then gave you a set of fixed assemblies and dependencies that were checked into source control. Without that, I don't see how this feature is usable.

**Comment by Wil73** [comment]
Follow me on this David.. it does NOT matter.  It does not matter if you can see how this would be usable.  This is not the province of engineers to see that.  They build what business asks for, that's it.  This is a stability vs featureset throughput decision.  Businesses make these decisions, not engineers.

This is CLEARLY a business need.  I have read literally over a HUNDRED posts of people asking for this.  So you build it.  You don't debate the sense of building to latest version. Some businesses want that aggressive strategy, some want build to oldest.  Nuget should let the business decide for themselves.

Allow your users to decide for themselves how aggressively they want to adopt new versions.  You build in the flexibility for the business to define this.  Engineers have no business making these sorts of decisions unilaterally for every one of their consumers.  This IS a business decision.

I will however leave you with this thought: if the oldest version were always the best choice, we would all still be running Windows 95...

**Comment by giggio** [comment]
I think you are being paternalistic. You should not decide how I build software. It is my decision, not yours.
What you are saying is "this is how I do it, and we should all do it this way". Please don't do that.
I do agree that in most projects people do not want that, as it may break things.
But in a cutting edge project maybe I want to be on the latest of everything, or latest minor, or patch. Maybe that is ok, because the project is just starting. Or maybe I want the latest patch version because it fixes bugs. I don't know. But please don't limit me.
I have used `^` in node without a lockfile, and very rarely did the build break. When the project reached a more stable version we added a lockfile. For some dependencies we removed the `^`. It was not a nightmare. And the node ecosystem changes much more rapidly than the .NET's ecosystem. Packages are released much more often.

**Comment by giggio** [comment]
Also, it would be very important to have an option to exclude prereleased versions.

**Comment by Wil73** [comment]
I am not saying that at all.  Are you guys not reading what I write? From day 1 I have argued that Nuget should be flexible enough to support each individual business strategy.  I am most certainly not being paternalistic here.  The problem is _Nuget assumes that resolving to oldest version is best_.

This is certainly not the case.  Nor is resolving to latest version always best.  What I am and have been saying is far more subtle than this.

We have David above arguing why oldest is best.  I really think this is a moot point:

_Each business must decide how conservative or aggressive their package adoption strategy ought to be._

Unfortunately Nuget does not offer a flexible resolution strategy here.  Under the covers Nuget assumes oldest is best for child dependencies.  This is a big problem for some companies who do not want oldest is best.

I am most certainly _not_ saying we should do things any one way. I am the one here arguing for a more flexible package adoption strategy.  A strategy that offers a package resolution strategy switch/toggle that the business/engineers configure for each project themselves.    Rather than have all this AI built under the covers that decides unilaterally how aggressive or conservative a business's package adoption strategy is we should be able to configure this ourselves.  How on earth can anyone unilaterally know how aggressive or conservative any given company's package adoption strategy ought to be?

Add a switch of some sort that allows Nuget to restore and resolve packages the way each business needs.  A customizable and configurable Nuget AI would be a huge gain.  Right now Nuget Resolution strategy is a black box and not configurable.  Nuget decides and we just have to live with it.

This has been a problem for six years now.  It's time to fix it.

**Comment by gyrolc** [comment]
Is there a timeline for this resolution? We have the exact same business need to always get the highest available version of our internal packages within a specified range.

**Comment by Wil73** [comment]
My biggest concern is that this will get dealt with, but the implementation will be inadequate.  I don't understand at all why this takes so long.  Just add a switch to nuget.exe to pass into the business logic.  How can this be that complicated?

I must be missing something.

**Comment by anangaur** [comment]
@Wil73 We typically publish the spec and announce it at the [NuGet/Announcements](https://github.com/nuget/announcements) repo for review. We respond to feedback and once reviewed, we begin the implementation. So please rest assured that we will take in your feedback :) Please subscribe to [NuGet/Announcements](https://github.com/nuget/announcements) to not miss out on the notifications.

@gyrolc This is on our backlog and we plan to work on this once we implement the features outlined on our [Roadmap blog post](https://blog.nuget.org/20170809/NuGet-Fall-2017-Roadmap.html). You can also keep track of the upcoming features by subscribing to the [NuGet/Announcements](https://github.com/nuget/announcements) repo.

**Comment by davidfowl** [comment]
> My biggest concern is that this will get dealt with, but the implementation will be inadequate. I don't understand at all why this takes so long. Just add a switch to nuget.exe to pass into the business logic. How can this be that complicated?

Why not send a pull request with the change then?

**Comment by Wil73** [comment]
That's a great idea David.  Why don't you send a pull request to implement the changes you are asking for?  I look forward to using them.  Thank you!

**Comment by springy76** [comment]
Regarding "old is safe":

> Everyone in the graph gets what they compiled against

Sorry this is not true: https://github.com/domaindrivendev/Swashbuckle.AspNetCore/issues/438#issuecomment-360066292

In that case it might be nice that Swashbuckle got its 1.0.4 StaticFiles it had been compiled against. But that StaticFiles gets NOTHING it got compiled against since 99,999% of the project already was at aspnetcore 2.0 at the time Swashbuckle got added and transitively pulled StaticFiles.

**Comment by davidfowl** [comment]
@springy76 the assessment of what caused that issue is incorrect. I clarified in the comment.

**Comment by springy76** [comment]
The biggest problem with this: No one tells you anything is wrong. You get a MissingMethodException sometime at runtime, maybe many hours or days after starting the app.

While transitive references may not be the direct cause for the problem the missing tooling for showing the availability of updates of transitive references **is**. With packages.config this would not have happened.

Or may I remember 2018-03-13: Some smart guy decided to delist all 2.0.0./2.0.1 aspnetcore NuGet packages due to CVE-2018-0787 which just blew up any package restore. Our build server just could not build the exactly same sources it had built successfully only some hours ago without having changed anything (but the NuGet servers did).
https://github.com/aspnet/Home/issues/2954#issuecomment-373064006

**Comment by markusschaber** [comment]
@springy76 That's precisely why one should **never ever** let his build servers build against a public repo.

Use a local repository for the CI systems, and restrict the build servers from accessing anything non-local. Apart from reproducible builds, this also helps with license compliance and security  @assessments, as you only have "known" packages with "known" licenses.

**Comment by abatishchev** [comment]
>you only have "known" packages with "known" licenses.

How so? Manual control of each package? You can do that either way. And to automate you need tooling 

>one should never ever let his build servers build against a public repo.

Oh, please. This is terrible burden on developers. Plus you again need tooling to sync local with public.

The only worse advice is to check-in packages under source control for same, illusional reasons

**Comment by markusschaber** [comment]
We need local-only builds anyways - OSS licenses is only one reasons, some other are:

- We need to be able to re-build older branches, so we can apply bug fixes to those branches with minimal side effects.
- For audits and certifications, builds must be reproducible, and we need to be able to proof that a deliverable resulted from a specific source.
- For our major products, everything (sources, toolchains, 3rd party modules) needed to build a release binary has to be deposited - some customers have contracts which grant them access to those artifacts in case we go out of business, or stop supporting those products.

Also, for OSS license compliance, every package has to be audited whether the license is clearly declared (surprisingly many packages come without license, or with contradicting or otherwise unclear legalese), and whether it's allowed in the context the package is used.

But some (exotic) licenses which restrict usage are just not acceptable for us or our customers, others like GPL are only acceptable in specific contexts, and others need specific caution, for example the "advertising clause" in the OpenSSL license results in some guidelines for our documentation and marketing teams.

Yes, that whole process is a high burden for developers in .NET core and even more in angular/npm land, but it's the only way to prevent ticking bombs for us and our customers. And, yes, we're using lots of tooling to assist developers.

**Comment by NightWatchman** [comment]
> Is there a timeline for this resolution? We have the exact same business need to always get the highest available version of our internal packages within a specified range.

I agree with this and hope to see a resolution soon for making the DependencyVersion property, which already works for package.config projects, work for projects that use the PackageReference method as well.

I see the value of "lowest is safest" when the Version is specified with something like `4.3.2` which, by default, resolves to >=4.3.2, but that argument goes right out the window if the user is specifying the version with something like `[3.0.22,3.0.24]`, or `[1.5.2,3)`. Clearly anybody who's bothering to go through the effort of specifying a version range that specific did so because they believe that the project works with any of the specified versions. Since it works with any of the specified versions, restricting it to the lowest possible version in the specification without any way of modifying that behavior is just really annoying.

**Comment by giggio** [comment]
Another 6 months went by and we got not response. This is a `priority:1` issue that is on the backlog for 3 years. Shouldn't this be at least planned for a release?

**Comment by mikenunan** [comment]
I'd like to add my voice to this, having just landed here after some frustrating time spent trying to specify a "latest" dependency in a nuspec. There absolutely are sane use cases for this, I can't quite believe it's not possible to simply use * to pull in a floating latest version, or 1.* or 1.3.* etc to restrict, or to use ranges somehow.

In my case I just need "latest" because the dep contains build-time tooling (that does some validation work) that I precisely _do_ want moving forward as new versions are released. Now this means I'm going to have to update the dependent components every time that tooling component changes.

**Comment by michaelTBF** [comment]
We want latest - if I specify 1.2.* I will always like the updates made to 1.2.* - nothing else makes sense.

Microsoft push Pipelines and CICD really much - but the basics behind the scenes are still way old.

I know you can make this better.

**Comment by omog61** [comment]
When I found Dr. ABAKA I was in desperate need of bringing my ex lover back. He left me for another woman. It happened so fast and I had no say in the situation at all. He just dumped me after 3 years with no explanation. I contact Dr.ABAKA through his website and He told me me what i need to do before he can help me and i did what he told me to, after i provided what he wanted, he cast a love spell to help us get back together. Shortly after he did his spell, my boyfriend started texting me again and felt horrible for what he just put me through. He said that I was the most important person in his life and he knows that now. We moved in together and he was more open to me than before and he started spending more time with me than before. Ever since Dr. ABAKA helped me, my partner is very stable, faithful and closer to me than before. I highly recommends Dr. ABAKA to anyone in need of help. Email: drabakaspelltemple@gmail.com, Call him or add him on whatsapp via: +2349063230051/https://drabakaspelltemple.blogspot.com

**Comment by omog61** [comment]
When I found Dr. ABAKA I was in desperate need of bringing my ex lover back. He left me for another woman. It happened so fast and I had no say in the situation at all. He just dumped me after 3 years with no explanation. I contact Dr.ABAKA through his website and He told me me what i need to do before he can help me and i did what he told me to, after i provided what he wanted, he cast a love spell to help us get back together. Shortly after he did his spell, my boyfriend started texting me again and felt horrible for what he just put me through. He said that I was the most important person in his life and he knows that now. We moved in together and he was more open to me than before and he started spending more time with me than before. Ever since Dr. ABAKA helped me, my partner is very stable, faithful and closer to me than before. I highly recommends Dr. ABAKA to anyone in need of help. Email: drabakaspelltemple@gmail.com, Call him or add him on whatsapp via: +2349063230051/https://drabakaspelltemple.blogspot.com

**Comment by GirishVilli** [comment]
Hi, Any update on this issue. Need a solution to control dependency resolution behavior for transitive dependencies

**Comment by TysonMN** [comment]
There was much talk above about first needing support for lock files when specifying packages (via `PackageReference` in project files).  This [feature was released](https://docs.microsoft.com/en-us/visualstudio/releasenotes/vs2017-relnotes#NuGet) on November 13, 2018 with Visual Studio 2017 version 15.9.0.  Here is a [direct link](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#locking-dependencies) to the feature documentation.

**Comment by anangaur** [comment]
This feature is on our backlog but not one of the top ones right now and yes its unblocked now since we implemented the lock file feature. 

@bender2k14, @GirishVilli can you elaborate a bit on your scenario even if its re-iteration of the reasons discussed above? Feel free to state it here or reach out to me via an email.

**Comment by TysonMN** [comment]
I don't have any long-term need for the `-DependencyVersion` option.  I was confused about how some aspects of NuGet package management work.

In the process of trying to remove binding redirects from my code, I thought a good solution was to [request `Serilog.AspNetCore` change their dependencies](https://github.com/serilog/serilog-aspnetcore/issues/85).  (I now know this was a bad idea.)  The [first response](https://github.com/serilog/serilog-aspnetcore/issues/85#issuecomment-481399598) I got was to try `Install-Package Serilog.AspNetCore -DependencyVersion Highest`.  But that didn't do anything different for any argument to `-DependencyVersion` (because, I know now, I am specifying my NuGet package dependencies via `PackageReference` tags in the project file...per the present issue).  When I [asked why](https://github.com/NuGet/Home/issues/8327) in this GitHub project, I was pointed to the present issue.

To help others like me, a [new issue](https://github.com/NuGet/Home/issues/8329) was created suggesting that a warning be given when `-DependencyVersion` is used with `PackageReference`-specified NuGet package dependencies (at least, until the present issue is resolved).

While I have the attention of a NuGet expert, here is one thing that I think might be helpful to me.  How can I "see" the NuGet dependency graph of my solution?  Specifically, if some project in my solution depends on the same NuGet package (potentially with different version constraints), I would like to know what all the paths in the NuGet dependency graph are to that NuGet package as well as the version constraint contributed by each of these paths.  The goal is to figure out a better way to handle duplicate indirect dependencies than a binding redirect (such as making this indirect dependency into a direct dependency and to picking the desired version).  Does that all seem reasonable to you?

My plan is to understand this well enough that I can give a talk a my company to help everyone improve their NuGet package dependency management.

**Comment by anangaur** [comment]
> While I have the attention of a NuGet expert, here is one thing that I think might be helpful to me. How can I "see" the NuGet dependency graph of my solution? ...

One of the ways is to run [`dotnet list --include-transitive`](https://docs.microsoft.com/en-us/dotnet/core/tools/dotnet-list-package#options). This funtionality is not yet available with Visual Studio, but we will get there.

The other part is to understand why a specific version was selected. We do not have a mechanism to show this information, but may be having a `why` command would help here. Alternatively, you can look at the [lock file](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#locking-dependencies) to figure out what was requested vs. what got resolved , but this requires parsing of the lock file.

**Comment by TannerBragg** [comment]
I don't understand the danger in allowing package mantainers to add additional options to their csproj files to specify the dependency resolution strategy.  At my workplace, minor version never break code and we always want to use the latest minor version.  If we need to lock to a specific version for a specific build, we would use the lock file.

Any traction on this issue?  It's outstanding.  It's desired.  Why are we being ignored?  Does someone have the CEO's twitter?  I hear bad publicity on Twitter gets fast responses.

**Comment by tebeco** [comment]
hello there 
i think i can't just repeat thing that are already on the thread
i have also issue for example with Serilog that has many Microsoft.Extensions.Xxx as transitive in 2.x being pulled down to aspnetcore3.1
i let you guess the disaster

(Some PR were rejected on Serilog to update to msft.extensions.xx 3.1.x)

it forces us to pin down transitive explicitly 
we are re-creating one of the biggest issue `package.config` had => having both direct and transitive pinned per project 

i would really love to see this land
as aspnetcore3.1 is now netcoreapp3.1
BUT depends on msft.exrension 3.1 that are netstandard2.0, unless all third part updates the default behavior will attempt to get 2.x as potential transitive in 3.1 apps

**Comment by davidfowl** [comment]
@TannerBragg do the things you depend on have the same guarantee?

**Comment by tebeco** [comment]
sorry about pushing that topic again, is there any info on how to to get that `Highest` resolution for transitive from an `MsBuild Props` today ?
especially when using `[*, )` and lockfile properly.

For example :
If you depend on `Serilog`
You would get transitive resolution of some `Microsoft.Extensions.xxxx` (like logging) to 2.x while the Host project is `aspnetcore / 
netcoreapp3.1`
It seems to happen because serilog also depends on `Microsoft.Extensions.xxx >= 2.1` but the resolution defaults to "Lowest" for example
As it is open on the right, i would love to have a switch in my `sources.props` to be able to choose how i want that resolution to happen so i can try whatever suits my needs

**Comment by TannerBragg** [comment]
> @TannerBragg do the things you depend on have the same guarantee?

I'll be honest with you.  I completely gave up on this one.  I took my nice modular assemblies and made big ones that hold all the source separated by folders and namespaced them appropriately.  It was the easiest solution over fighting with the tool.

**Comment by AgSync-Aaron** [comment]
First, let me say I have nothing to add to this conversation. I would just like to be able to use `Install-Package -DependencyVersion Highest [package_id]` for .NET Core projects. I've been manually installing dependencies when I want to force the latest version and that's a PITA. Thanks for the work you do. I just wish this issue was on the list.

**Comment by springy76** [comment]
@tebeco The pity is that between most `Microsoft.Extensions.*.Abstracts` v2.x and v3.x there is no single breaking change, so in SemVer-thinking there should be no major number change at all.
But on the other hand I don't want different NuGet versions of netcore parts so I can use a single build variable to update them all, for example from 3.1.4 to 3.1.5.

**Comment by TysonMN** [comment]
You are applying the converse of [SemVer rule 8](https://semver.org/#spec-item-8) instead of rule 8 itself.  SemVer allows the major version to be incremented for any reason (as long as the minor and patch versions are reset to 0.)

For example, I think it would be appropriate to increment the major version level if the code was totally rewritten (maybe even in another language) while keeping the API (and even all behavior) unchanged.

**Comment by reiseder** [comment]
Are there any updates on this issue?

It would save a ton of work for releasing well maintained internal packages at my company.

**Comment by tebeco** [comment]
I can only suggest anyone to upvote the first comment / issue in this issue, as it seems to be the relevant metrics being used to prioritize how much user want something to move on
activity is hard to quantify by bot (because spam / irrelevant etc ...) while a +1 is quite clear

**Comment by ChaseFlorell** [comment]
I mean, 4 years old, perhaps it's just not a priority.

**Comment by markusschaber** [comment]
Is the NuGet team open for pull requests?

**Comment by rdeago** [comment]
> Is the NuGet team open for pull requests?

Looks like they are.
https://github.com/NuGet/Home/blob/dev/CONTRIBUTING.md

**Comment by IanKemp** [comment]
Still waiting for this critical feature, Microsoft!

**Comment by mikamins** [comment]
If there's no way to override the lowest version behavior, how are people expected to deal with security vulnerabilities in transitive dependencies?

As a concrete example, [Azure.Core](https://www.nuget.org/packages/Azure.Core/1.11.0) has a dependency on System.Text.Encodings.Web 4.6.0 which is [susceptible to remote code execution](https://github.com/dotnet/announcements/issues/178). Even though the issue was fixed in System.Text.Encodings.Web 4.7.2 almost a year ago, I get the vulnerable version if I consume the latest Azure.Core.

**Comment by RandyT97** [comment]
> If there's no way to override the lowest version behavior, how are people expected to deal with security vulnerabilities in transitive dependencies?
> 
> As a concrete example, [Azure.Core](https://www.nuget.org/packages/Azure.Core/1.11.0) has a dependency on System.Text.Encodings.Web 4.6.0 which is [susceptible to remote code execution](https://github.com/dotnet/announcements/issues/178). Even though the issue was fixed in System.Text.Encodings.Web 4.7.2 almost a year ago, I get the vulnerable version if I consume the latest Azure.Core.

Hi @mikamins , I actually ran into the same issue with specifically :
    "Microsoft.ServiceFabric.AspNetCore.Kestrel 4.2.477" => "Microsoft.ServiceFabric.AspNetCore.Abstractions 4.2.477" => "Microsoft.AspNetCore.Http.Abstractions 1.0.0"=>"System.Text.Encodings.Web 4.0.0":

I resolved this by adding System.Text.Encodings.Web 4.7.2 as a direct dependency of the msbuild project, and it will cause all lower level dependencies to reference v4.7.2 instead of v4.0.0

**Comment by rdeago** [comment]
@RandyT97 while your workaround of course works, it solves the problem for a specific package. Your project could well have other transitive dependencies suffering from the same problem without you even being aware of it.

Besides, a few months from now, anyone working on the project (or even some automatic tool) may remove the direct dependency on `System.Text.Encodings.Web` because it is not used directly, everything builds just fine without it, so why have it around?

**Comment by springy76** [comment]
net5 now totally f*cked up it just from the start. Being now at 5.0.4 some newest packages are at 5.0.3, 5.0.2, 5.0.1, 5.0.0, and some other versions.
And it doesn't matter whether I use 5.0.4 packages on 5.0.0 runtime or 5.0.0 packages on 5.0.4 runtime. All combinations just work and I can't tell the build server that this all probably is just wrong and smelly bullshit.

**Comment by HugoLemieux** [comment]
A lock file has been introduced in order to properly do this with PackageReference.
Are there any updates on this issue?

**Comment by canfucius** [comment]
So is this issue fixed? And will there be documentation to show how to properly handle float versioning throughout an entire project?

**Comment by tebeco** [comment]
the title is about direct OR TRANSITIVE
so it's still not possible no

also you can't decide about a strategy AT ALL, you have to go manually edit every single dependency one by one as a text file and replace digit with `*`
like when you use `dotnet add package` it won't care about other package using wildcard it will put the full version down

as for the lockfile, did they fixed
* the cross-platform issue raising error when you generate a lockfile on windows and use it on linux (or the opposite way)
* how does it now handle the SDK changing/adding on the fly reference ?

**Comment by Flavien** [comment]
Can we hear from the NuGet team? Is this:

- Currently being worked on (if so, what milestone?)
- On the roadmap but not a priority
- Not on the roadmap at all, if so, why?

**Comment by JonDouglas** [comment]
Hi all,

I'm going to write a message here for now because this issue hasn't had a reply from the team in quite some time. I apologize in advance.

This issue is currently the 3rd 👍  most upvoted issue in our client repository. There are many considerations to take into account for the behavior of [package resolution](https://docs.microsoft.com/en-us/nuget/concepts/dependency-resolution#dependency-resolution-rules) that have already been discussed amongst everyone here and [known workarounds & features](https://docs.microsoft.com/en-us/nuget/concepts/security-best-practices#manage-your-dependencies) that help you be more deterministic when resolving packages(especially with regards to security).

I want to be transparent with everyone and tell you that from my perspective(PM on the NuGet team), this issue is important but we currently do not have a plan for it. What that means is that we believe there are other critical work to complete this year prior to getting time to invest into this (and the consideration of capacity). **This can change however.** 

We also need to understand this issue further, the current pain points, and know what is possible for the product given many of the previous blockers have been implemented or have plans to be(i.e. lock files / package overriding / etc) and the limitations we have. 

Your feedback on how we can move forward with this issue and your ideal happy paths in your daily management of dependencies have already helped us tremendously. Please feel free to keep commenting on your ideal experiences so we can make the best solution. Knowing the desired end goal/experience will keep us headed in the right direction.

I am currently working with our teams to produce a year's living roadmap to clue everyone in on what we have planned for 2022. When we publish this, please provide your feedback so we can understand how this issue weighs in comparison to other highly requested features. This will help us get our plans right and be more community driven. 

Thank you.

**Comment by AgSync-Aaron** [comment]
Thank you for providing an update, even if it is unfortunate news.

**Comment by Flavien** [comment]
Thanks, appreciate the update, and glad to know it's at least (somewhat) on your radar.

**Comment by HugoLemieux** [comment]
The ideal experience would be to resolve the transitive dependencies with a defined resolution strategy and consolidate the transitive dependencies for a solution because the lock file is by project.

**Comment by JonDouglas** [comment]
@HugoLemieux In the meantime, how would that look in your current development experience? What type of CLI flags, VS UI, or other happy paths would allow you to do that?

Consider the ability to: 

1. Decide the dependency resolution strategy for both direct & transitive levels.
2. Override or set an exact version for transitive packages.

How would you like to express that in your configurations today?

**Comment by HugoLemieux** [comment]
It would be great to have the same experience as it was with the packages.config.

In order to resolve the transitive dependencies with a defined resolution strategy:
- CLI flag: -DependencyVersion [Lowest, HighestPatch, HighestMinor, Highest, Ignore]
- VS UI:
![image](https://user-images.githubusercontent.com/12341715/148575642-62e23b25-307b-483b-8655-fee1545aea1b.png)

The VS UI in order to consolidate all the dependencies (direct and transitive) for a solution:
![image](https://user-images.githubusercontent.com/12341715/148576810-d6762544-c252-41c7-9219-2a3c3d873ea8.png)

I think these could work using lock file by project.

**Comment by nicolascotton** [comment]
We would like to have better control and cohesion between the versions of our dependencies (transitive included) across multiple projects (workspace/solution).
This could be done with a global lock file and a granular control of the transitive dependencies within.

**Comment by idg10** [comment]
One particular scenario that causes us to want something like this might be worth calling out as a special case: it's the fact that every single package in `Microsoft.Extensions.*` revs every single time there's an update to .NET whether anything has changed in a package or not. We have 40 odd open source libraries that we maintain, and every week or so we get a deluge of Dependabot notifications telling us about the zillion places in which it thinks we should update one of these packages.

So what we'd really like is some way to express the policy that we always want the latest available patch of those components by default. The combination of Microsoft's high standards for backwards compatibility plus the fact that most of the time a new patch version on one of these components hasn't changed anything of substance means that we basically always want to take the latest version of these, but tend to want to be deliberate about everything else.

(Alternatively, if these packages would only drop new patch versions when something has actually changed, that would be much better. I wouldn't be drowning in dependabot requests, and I'd also know when there is actually a meaningful change.)

**Comment by viceice** [comment]
Would be nice to have the `dependencyVersion `[^1] option again, so every body can decide which updated to get

[^1]: https://learn.microsoft.com/en-us/nuget/reference/nuget-config-file#config-section

**Comment by philipborg** [comment]
My issue is in large part that there is no pretty way to solve outdated transient dependencies, especially when there is a critical security issue which is often. If I want to update the transient dependency I currently need to add it as a direct dependency, even though I have no actual direct dependency on it. This both bloats the project file, confuses the hell out of any project readers and observations of dependency trees either using tools or lock files.

I would prefer to be able to set both default transient dependency resolution and/or (preferably and) override transient dependencies in a clearer manner which doesn't add it as a direct dependency. Ideally for specific transient dependencies.

**Comment by tdhintz** [comment]
It's crazy that there is no plan articulated for this problem.  Our vulnerability scanners check dependencies and report problems on dependencies we don't even deploy.  Manually managing these alerts is not practical.  Automation is needed.

**Comment by AgSync-Aaron** [comment]
> It's crazy that there is no plan articulated for this problem. Our vulnerability scanners check dependencies and report problems on dependencies we don't even deploy. Manually managing these alerts is not practical. Automation is needed.

Take a hintz! Microsoft will never fix this issue until there is some bad press about it. Microsoft is being apathetic to the risk they are putting their own customers in, in regard to critical vulnerabilities. 

Maybe this project can become mainstream... https://fsprojects.github.io/Paket/

**Comment by davidfowl** [comment]
I just re-read this, and there's been lots of changes to nuget since this was filed ~5 years ago. Have the problems changed? There are now lock files and central package management that allows you to control the resolution of packages whether they are transitive or not. 

I don't want to come off as being "paternalistic" and I want developers to have control, but package dependency resolution requires a global view of the world (the world being your project's transitive closure). Packet had to work around many assumptions in how package versioning works, and some strategies will never work with certain packages. 

What might help move this forward are examples of problems and existing workarounds. That would help narrow down how the existing tools have gaps. Some questions on the top of my head:

1. Where do you specify strategy, and where would it be persisted?
2. How do you specify this for transitive deps and where is it persisted in that case?
3. What's the granularity of specifying this strategy?
4. When does NuGet decide to check for updates? This probably needs to be an explicit gesture (we must be solving this problem today).
5. Does this mode require a lock file? If not, I guess the assets file can be used in place while it exists?
6. Is this something that is only expressed when consuming packages or also when building them? Or is it both?

**Comment by giggio** [comment]
@davidfowl 

1. .csproj (npm style which is already well understood, with `~` and `^`)
2. through command line, persisted in lock file
3. can you elaborate? what do you mean?
4. NuGet never updates anything until the user requests it, for restores it uses the information it has from .csproj and a lock file, if it exists
5. no, but it can help, see item 2
6. I'm not sure about this one, but probably both, so behavior is consistent between uses

**Comment by tdhintz** [comment]
@davidfowl 

1. We specify a NuGet config in the build pipeline with the specification we want.
2. The NuGet configuration file referenced through the command line in the build YML.
3. My desire is that the transitive policy applies to all transitives, regardless the source.
4. When I run the restore command (via the YML).
5. ?
6. When the restore is run using the NuGet configuration I specify.

**Comment by mcm-ham** [comment]
> It's crazy that there is no plan articulated for this problem. Our vulnerability scanners check dependencies and report problems on dependencies we don't even deploy. Manually managing these alerts is not practical. Automation is needed.

I'm in the same boat, however there are other approaches that could be implemented for this specific concern, I've suggested these in a separate ticket https://github.com/NuGet/Home/issues/12341

**Comment by clyvari** [comment]
I stumbled upon this discussion looking to do exactly as the OP stated, however reading the discussion and thinking long and hard about it, I definitely changed my mind: I don't think this feature is really needed nor desirable as stated, but I think some tweaks to the current features could help.

Keep in mind I'm talking only about transitive packages, since we have the ability, to specify `Version=*` in `*.csproj` files for direct packages. It's about the ability to configure one way or the other the resolution for _transitive_ packages of a dependency you have.

### Security
I think relying on "highest possible" dependency version resolution for security is a terrible idea, since there is absolutely no guarantee you will get a version without security issues. And assuming we could specify `*` for a transitive dependency version, there is no guarantee you will get the latest version anyway.
So that would be something that _makes you think_ you are getting the latest and greatest patched versions of everything, when in fact you do not. The security argument is, in my opinion irrelevant.

--

Another point: we can classify packages in two types.
* The purely technical ones (Newtonsoft, ...) that are "algorithms-only", they are self contained.
* The packages that interact with the outside world: A nuget package to communicate with an Azure API for instance.

#### Self-contained packages
I don't think there is a need to change anything. As long as they work, they work, even if the version used is "ancient". And if you do rely on specific optimizations or features _of a transitive package_, well it should definitely be part of your direct project dependencies.

#### External resources packages
There is usually NO good reason to use an older version of those packages. Most of the time, you do want the latest package possible. There is usually only 1 instance of a Web API you want to consume, that definitely works with the latest packages, and most definitely do not work with older packages.
And for those packages, even tough the code interface did not change, it is entirely possible that internal changes produce different Requests to the External API: Even though your code can compile AND run without exceptions, it might not work the the resource you want to consume.

So those packages sound like a good candidate for the feature. However, those packages are well identified, you definitely know which resources you are consuming (or else you have more pressing concerns), and can already use Central Package Management [Transitive Pinning](https://learn.microsoft.com/en-us/nuget/consume-packages/central-package-management#transitive-pinning) to tell which version you want (but you cannot at the moment specify `Version=*` in central package management, which is one of my suggestions).
But most likely, they should be included as direct dependecies of your poject, see suggestion 2).

TL;DR: I think we only really need custom _transitive_ dependency resolution for a small well known subset of packages, allowing global "Highest"-version for transitive packages will cause more trouble than it is worth.

--

So then, a couple of suggestions:
### 1. Allow floating versions in Central Package Management (CPM)
Basically allow the exact syntax we have in `*.csproj` at the moment. Using floating versions at the moment result in error [NU1011](https://learn.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu1011).
Maybe only allow floating versions when `packages.lock.json` is enabled. CPM does work with package lock, so I don't see any blocker.

### 2. Add a  `.nuspec` option to mark some dependencies as "direct"
It helps solves the issue for packages that depend on an "API nuget".

> Exemple: An helper library (`Lib1`) that simplifies working with an Azure API. This helper library has a dependency on an Azure API Nuget (`Azure.X.nupkg`).

As a Nuget publisher, my package (`Lib1.nupkg`) works for the foreseeable future, with version >= 2.0.0 of `Azure.X.nupkg`. I would like to mark this dependency as "direct", so a user of my helper library has more control over this dependency.
As consumer of `Lib1`, I want to be able to easily control the  `Azure.X` dependency and update it when needed.

With the new `.nuspec` "direct" dependency option, as a consumer of `Lib1`, adding `Lib1.nupkg` to my project would add `Lib1` as a direct reference, but also `Azure.X` as a direct dependency, instead of a transitive one.

My `*.csproj` would look like this after adding only `Lib1` :
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <!-- ... -->
  <ItemGroup>
    <PackageVersion Include="Lib1" Version="1.0.0" />
    <!-- Below was added automatically, this dependency is less hidden and I can easily mark it as floating if I wish to. -->
    <!-- "direct" dependencies in Azure.X package (if any) should recursively follow the same logic, and be included here -->
    <!-- Since it's not a transitive dependency, I will easily be able to update this package -->
    <PackageVersion Include="Azure.X" Version "2.0.0" />
  </ItemGroup>
</Project>
```
Nuget that should be referenced directly because they are "important" to your project can now be automatically included.
When there is an update to this package, I will be prompted to update this dependency in the future.

### 3. Warn if my code use transitive package code
I'll keep this one short since it's not related to Nuget directly: I should not be able to use code existing only in a transitive dependency without at least a warning, hinting me at including the package I'm using as a direct dependency.

--

Having thought about all the use cases I have met, I think those first 2 suggestions would cover pretty much everything, and allow me to be pretty confident in the management of my transitive packages.

**Comment by voroninp** [comment]
@clyvari If this strategy can be set individually for each dependency/package source, I'd turn it on for packages in my private feed, so whenever  `minor` or `patch` get bump, packages are automatically updated.

**Comment by clyvari** [comment]
@voroninp That's already possible for direct dependecies of your projet. If you add a depency with `Version=2.*`, you will automatically use the latest version possible that is <3.0.
It's only an issue for transitive dependencies, when you add a nuget `Lib1` that itself uses, say, `System.Text.Json`. You don't really care what version of System.Text.Json `Lib1` uses, as long as it works. And if you do, well that means you care enough about `System.Text.Json` to add it as a direct dependency of your project.

**Comment by voroninp** [comment]
@clyvari , let's replace System.Text.Json with MyLib and imagine larger (deeper) dependency graph. 

Now imagine, we found a bug in MyLib.

How many manipulations/rebuilds are required to update end project with the fixed MyLib depending on resolution strategy?

**Comment by clyvari** [comment]
@voroninp Assuming a bug in any dependency, at the moment you can (and probably should) use CPM with Transitive Pinning.
It's 1 commit to the `Dependency.Packages.props` file, and 1 build.

If at the root of a solution, it will be common to all projects in a solution, useful if the dependency is used in multiple projects.

Here is why using "highest possible" resolution strategy isn't sufficient :
Let's assume you have a several packages referencing VulnerableLib.
One requires `VulnerableLib Version=[5.2, )`, another `VulnerableLib [5.0, 5.7)`.
Latest `VulnerableLib` version is 5.7
You specify "Highest possible version" as a resolution strategy.

The version that satisfy your requirement is `5.6`.
Let's say a bug is found in VulnerableLib 5.x versions, fixed in version 5.8

Despite your "Highest possible version" as a resolution strategy, `VulnerableLib` will still resolve to vulnerable version 5.6.

And worse, because it is not intuitive (the version requirements might be deep down the dependency tree), you might be mislead into thinking: "VulnerableLib was bumped, all my projects use "Highest possible version" for transitive packages, I'm safe."

**Comment by tdhintz** [comment]
Its true that “highest possible” doesn’t always fix the vulnerability, but in my experience it fixes most.  An audit report will tell you if it didn’t fix something and enables you to focus on the less easily resolved issues.

**Comment by voroninp** [comment]
@clyvari > another VulnerableLib [5.0, 5.7)

I'd question first this range ;-)

**Comment by seanamos** [comment]
I never understood why Nuget had to re-invent the wheel here and come up with this bonkers system that we now sit with.

It's confusing for users coming from anywhere else and almost ensures you are going to have security vulnerabilities, even though patches are available. It's also encouraged an ecosystem where only a minimum version constraint is specified on dependencies, MS packages being a prime example of this.

Pre-dating Nuget, package managers had some simple concepts:

1. There is a lock file for repeatability.
2. Packages specify their dependencies.
3. Those dependencies have a minimum and a maximum version constraint.
4. Maximum compatible versions are chosen by default.

**Comment by springy76** [comment]
NuGet developed (or maybe always was) into a new kind of DLL hell - source of random "Method not found" exceptions at runtime (can't the exe-compiler or at least the publish task just check everything once?) 

I'm just right now adding the net6 package `ErikEJ.EntityFrameworkCore.SqlServer.DateOnlyTimeOnly` to a net7 project which already uses EFcore.SqlServer 7.0.x, the package itself only to extend EFcore.SqlServer 6.0.0+

Thats what "Preview Changes" tells me:

```txt
Uninstalling:

Microsoft.Win32.Registry.5.0.0
System.Globalization.4.3.0
System.IO.4.3.0
System.Reflection.4.3.0
System.Reflection.Primitives.4.3.0
System.Resources.ResourceManager.4.3.0
System.Threading.Tasks.4.3.0

Updates:

Azure.Core.1.24.0 -> Azure.Core.1.25.0
Azure.Identity.1.6.0 -> Azure.Identity.1.7.0
Microsoft.Data.SqlClient.5.0.1 -> Microsoft.Data.SqlClient.5.1.0
Microsoft.Data.SqlClient.SNI.runtime.5.0.1 -> Microsoft.Data.SqlClient.SNI.runtime.5.1.0
Microsoft.Identity.Client.4.45.0 -> Microsoft.Identity.Client.4.47.2
Microsoft.IdentityModel.Abstractions.6.21.0 -> Microsoft.IdentityModel.Abstractions.6.24.0
Microsoft.IdentityModel.JsonWebTokens.6.21.0 -> Microsoft.IdentityModel.JsonWebTokens.6.24.0
Microsoft.IdentityModel.Logging.6.21.0 -> Microsoft.IdentityModel.Logging.6.24.0
Microsoft.IdentityModel.Protocols.6.21.0 -> Microsoft.IdentityModel.Protocols.6.24.0
Microsoft.IdentityModel.Protocols.OpenIdConnect.6.21.0 -> Microsoft.IdentityModel.Protocols.OpenIdConnect.6.24.0
Microsoft.IdentityModel.Tokens.6.21.0 -> Microsoft.IdentityModel.Tokens.6.24.0
Microsoft.NETCore.Platforms.5.0.0 -> Microsoft.NETCore.Platforms.1.1.0
Microsoft.Win32.SystemEvents.5.0.0 -> Microsoft.Win32.SystemEvents.6.0.0
System.Buffers.4.5.1 -> System.Buffers.4.5.0
System.Configuration.ConfigurationManager.5.0.0 -> System.Configuration.ConfigurationManager.6.0.1
System.Diagnostics.DiagnosticSource.5.0.0 -> System.Diagnostics.DiagnosticSource.6.0.0
System.Drawing.Common.5.0.0 -> System.Drawing.Common.6.0.0
System.IdentityModel.Tokens.Jwt.6.21.0 -> System.IdentityModel.Tokens.Jwt.6.24.0
System.Runtime.Caching.5.0.0 -> System.Runtime.Caching.6.0.0
System.Security.AccessControl.5.0.0 -> System.Security.AccessControl.6.0.0
System.Security.Cryptography.ProtectedData.5.0.0 -> System.Security.Cryptography.ProtectedData.6.0.0
System.Security.Permissions.5.0.0 -> System.Security.Permissions.6.0.0
System.Text.Encoding.CodePages.5.0.0 -> System.Text.Encoding.CodePages.6.0.0
System.Windows.Extensions.5.0.0 -> System.Windows.Extensions.6.0.0

Installing:

ErikEJ.EntityFrameworkCore.SqlServer.DateOnlyTimeOnly.7.0.1
System.Runtime.CompilerServices.Unsafe.6.0.0
```

It's a wild mix of up- and downgrades, outtake:
```txt
Microsoft.NETCore.Platforms.5.0.0 -> Microsoft.NETCore.Platforms.1.1.0
Microsoft.Win32.SystemEvents.5.0.0 -> Microsoft.Win32.SystemEvents.6.0.0
System.Buffers.4.5.1 -> System.Buffers.4.5.0
```

I can just only hope this has no serious effect on the final app.

**Comment by yanivru** [comment]
In my company, we develop several nuget packages and consume them in other programs.
E.g. P1 <- N1 <- E1
Program1 (P1) references internal nuget N1 that references some external nuget (E1). When I fix bugs in N1 we increase the patch part of the version.

P1 references N1 with floating version (e.g. 1.*). So that P1 (and all other Ps) will take all the bug fixes automatically in next build.

Now if a newer version of E1 was released (e.g. 1.0.1), we want P1 to get it (cause it's might contain a bug fix, or security fix). What we usually do is update N1 to reference the newer N1 (1.0.1), otherwise, we will need to go over all the consuming projects (P1, P2, …) and reference E1 (1.0.1) directly, which is ugly and time consuming.

But since P1 might be directly refencing older version of E1 (1.0.0) it might cause a warning of package downgrade, and because treat warnings as errors it breaks P1.

So eventually the change to N1 is a breaking change. Which is not ideal. Either we break some builds or   we increase major of N1 each time we change some minor version of some referenced nuget package (time consuming cause all P's need to be changed). 

Is there a better way to do it, that will solve both problems?

What I would like to do is that N1 references the minimum version it requires (e.g. E1 1.0.0) and when P1 is compiled it takes the highest E1, with the bug fixes. This way there won't be a package downgrade error and still the fixes will arrive to all refencing programs (Px). This is not feasible currently since there is no "Biggest major" dependency resolution.

 We don't really care about deterministic builds. Our builds are not deterministic anyway, since we use floating versions and for other reasons.

**Comment by brentthierens** [comment]
Long time no response anymore, any update on this? I really need to be able to use the latest minor version of dependencies without having to recompile all other packages

**Comment by Madajevas** [comment]
Is this even being worked on?

**Comment by giggio** [comment]
> Is this even being worked on?

@Madajevas It doesn't seem that this is even being discussed, they just released their roadmap for .NET 9 and it is not in it. See #13143.

**Comment by supergibbs** [comment]
Pretty sure similar examples have been made but here is another one:

I was addressing [CVE-2024-21319](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2024-21319) identified in `System.IdentityModel.Tokens.Jwt:6.24.0`. I did not use this directly but I was using latest version of `Microsoft.EntityFrameworkCore.SqlServer:8.0.2` which brought in `Microsoft.Data.SqlClient:5.1.4` -> `Microsoft.IdentityModel.Protocols.OpenIdConnect:6.24.0` and then `System.IdentityModel.Tokens.Jwt:6.24.0`

My code was working fine so to fix this, I explicitly brought in `Microsoft.Data.SqlClient:5.2.0`.

This is a simple example but imagine there are more. Now I am having to manage versions of dependencies for dependencies and at scale I can see this causing conflicts and a lot of work to get it right. I don't want to specify every transitive package explicitly and I don't want to miss out on minor fixes and improvements. If I do specify, I run the risk of it not actually being needed if whatever library used it changes. I'd prefer to use the latest patch or even minor version automatically and if there are issues, explicitly "downgrade" as needed. Like maybe `5.3.0` comes out and breaks `Microsoft.EntityFrameworkCore.SqlServer`, I can specify `Microsoft.Data.SqlClient:5.2.*` instead.

Lots of opinions here, I think everyone would be happy if we had the option though. For fun, run: `dotnet list package --outdated --include-transitive` on a large solution and see how much is out of date even if your direct dependencies aren't.

**Comment by JonDouglas** [comment]
Hello,

I’m proposing a solution to this issue by introducing an opt-in mode for SemVer compatibility, which would align NuGet with the default behaviors of other modern package managers. This approach would allow developers to choose whether they want to enable SemVer-compatible resolution for both top-level and transitive dependencies.

I've created a new issue to discuss this proposal in more detail:
https://github.com/NuGet/Home/issues/13779

Please feel free to add your comments and upvote over there. Since this issue has over 100 comments and has been open for a while, shifting the conversation to a new issue will help us focus on whether this feature makes sense and how it could resolve the original challenges brought up here.

Do also check out some of our docs on this topic here which provides guidance today:

https://learn.microsoft.com/nuget/concepts/auditing-packages

https://github.com/NuGet/docs.microsoft.com-nuget/pull/3336/files

**Comment by Matthias-Heinz** [comment]
> Pretty sure similar examples have been made but here is another one:
> 
> I was addressing [CVE-2024-21319](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2024-21319) identified in `System.IdentityModel.Tokens.Jwt:6.24.0`. I did not use this directly but I was using latest version of `Microsoft.EntityFrameworkCore.SqlServer:8.0.2` which brought in `Microsoft.Data.SqlClient:5.1.4` -> `Microsoft.IdentityModel.Protocols.OpenIdConnect:6.24.0` and then `System.IdentityModel.Tokens.Jwt:6.24.0`
> 
> My code was working fine so to fix this, I explicitly brought in `Microsoft.Data.SqlClient:5.2.0`.
> 
> This is a simple example but imagine there are more. Now I am having to manage versions of dependencies for dependencies and at scale I can see this causing conflicts and a lot of work to get it right. I don't want to specify every transitive package explicitly and I don't want to miss out on minor fixes and improvements. If I do specify, I run the risk of it not actually being needed if whatever library used it changes. I'd prefer to use the latest patch or even minor version automatically and if there are issues, explicitly "downgrade" as needed. Like maybe `5.3.0` comes out and breaks `Microsoft.EntityFrameworkCore.SqlServer`, I can specify `Microsoft.Data.SqlClient:5.2.*` instead.
> 
> Lots of opinions here, I think everyone would be happy if we had the option though. For fun, run: `dotnet list package --outdated --include-transitive` on a large solution and see how much is out of date even if your direct dependencies aren't.

Yeah, I totally agree: We need to be able to specify **DependencyVersion: Highest** again.

The current way leaves developers with a total clusterf*** of transient dependencies. This is a screenshot of **some** of the versions pulled by Nuget Package Manager for *System.Text.Json* for our solution/project (.NET 8 project) at this point in time. We don't use any Top-Level dependency for that particular package.

![Image](https://github.com/user-attachments/assets/7eabfe60-c9ca-4012-beac-4a6fddc1f149)

As you can see, there's even versions with known vulnerabilities installed due to that fact. I'd like to be able to at least get those out of the system! Preferably without having to check for vulnerabilities regularly and installing said transient dependencies manually!

**Comment by ceztko** [comment]
I'm not 100% sure about what's asked in this issue: are you suggesting the version resolution algorithm to honor strings like `[2.3.0,2.5.*)` (which is invalid today), which it could mean bind to at least 2.3.0, but resolve to the highest in 2.5.*? Or by other means be able to describe such constraint? I noticed https://github.com/NuGet/Home/issues/1192, https://github.com/NuGet/Home/issues/4789 were closed, either by marking as duplicate or without fully addressing what asked. Those issues had more straight to the point description title, IMO, which helps in triaging/prioritizing the request.

**Comment by odinserj** [comment]
### How It Worked Before

Previously it was possible to specify the [`-DependencyVersion`](https://learn.microsoft.com/en-us/nuget/reference/cli-reference/cli-ref-install#options) flag when installing a package, please see the example below. This ensures that all the dependent packages used by the `Microsoft.EntityFrameworkCore.SqlServer ` package will be installed using the highest supported version.

```powershell
nuget.exe install Microsoft.EntityFrameworkCore.SqlServer -DependencyVersion Highest
```

For example, this package depends on the `Microsoft.Data.SqlClient` package. By default, the 5.1.6 version will be installed as the one specified in the Microsoft.EntityFrameworkCore.SqlServer dependencies. And with the `Highest` option, we'll get the most fresh one instead, 6.0.1 that contains great new features. And with the `HighestMinor` option, we'll get version 5.2.2 that contains a number of bug fixes, comparing to 5.1.6 (in case we are afraid of breaking changes).

So, as a developer that uses Microsoft.EntityFrameworkCore.SqlServer, I have a number of options to specify the behavior I would like to see in my application when dealing with dependencies.

The problem is that after switching to `PackageReference` we don't have any way to specify this behavior, and all the dependencies are installed using the most compatible `Lowest` option. So, all the dependencies are at least installed, but I will be unable to use latest fixes or features available.

### Suggested Solution

So I would like to see the following scenario supported when using the `PackageReference` tag:

```xml
<PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.1" DependencyVersion="HighestMinor" />
```

or if I risky enough:

```xml
<PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="*" DependencyVersion="Highest" />
```

or if I prefer compatibility (behavior is equal to the default one):

```xml
<PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.1" DependencyVersion="Lowest" />
```

This feature is already implemented by the NuGet client itself, and it already works. The only problem is that the `PackageReference` tag doesn't allow to use it.

**Comment by swettstein** [comment]
8 years and they stubbornly refuses to give people what they want. it works with Central Package Management but i wish it also worked directly with Package Reference.

**Comment by voroninp** [comment]
@swettstein , 

> it works with Central Package Management

huh? Did I miss something?

**Comment by swettstein** [comment]
it's not an exact fix but a workaround.

in my situation, i manage a lot of internal packages on the platform team and i want to push updates to dependencies that may not be direct dependencies. our dev teams' solutions typically reference the direct dependency with a wildcard for the patch version. currently, if i want to push a patch to a base dependency, i have to also bump the patch version of the direct dependency. it's not ideal and Nuget refuses to allow a better way for those who so choose.

example situation: DirectDependency depends on BaseDependency with version range `[2.0.0,2.1.0)`. Published versions of BaseDependency: 2.0.0, 2.0.1, 2.0.2. Published versions of DirectDependency: 2.0.0, 2.0.1.

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <GeneratePackageOnBuild>True</GeneratePackageOnBuild>
  </PropertyGroup>  
  <ItemGroup>
    <PackageReference Include="BaseDependency" Version="[2.0.0,2.1.0)" />
  </ItemGroup>
</Project>
```

with Nuget, the consuming project gets DirectDependency 2.0.1 and BaseDependency 2.0.1 (since when 2.0.1 DirectDependency was built, it resolved BaseDependency at 2.0.1). even when i push BaseDependency 2.0.2, the consuming project will not resolve it since it uses the lowest possible version.

with CPM, i can push a default Directory.Packages.props file like the one below:
```xml
<Project>
  <PropertyGroup>
    <ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
    <CentralPackageTransitivePinningEnabled>true</CentralPackageTransitivePinningEnabled>
    <CentralPackageFloatingVersionsEnabled>true</CentralPackageFloatingVersionsEnabled>
  </PropertyGroup>
  <ItemGroup>
    <PackageVersion Include="DirectDependency" Version="2.0.*" />
    <PackageVersion Include="BaseDependency" Version="2.0.*" />
  </ItemGroup>
</Project>
```

dev teams can reference the DirectDependency in their projects:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="DirectDependency" />
  </ItemGroup>
</Project>
```

now when i push BaseDependency 2.0.2, the consuming project will automatically pick it up. and to be flexible, overrides are always possible, e.g. a team chooses to not pick up new patches and instead leverage something like dependabot.

ideally this would rather be something like this (and i wouldn't have to worry about pushing a Directory.Packages.props file):
```xml
<PackageReference Include="DirectDependency" Version="2.0.*" DependencyVersion="Highest" />
```

**Comment by Grastveit** [comment]
Arrived at the same conclusion independently, from the transitive-bump-in-an-internal-feed angle.

+1 on the oldest-newest axis: That's what #14919 was for, duped here.

Filed #14917 for a separate but related axis: scoped re-resolution (`--update-package <pkg>` on restore, re-resolve one package, leave the rest of the lock file alone). Backed by analysis and prior art from Cargo, Bundler, npm, pip-tools, and Go, all of which ship the equivalent. Would also make the existing #5553 hotfix-workaround (pinning a transitive as a direct `PackageReference`) less painful with lock files enabled.
